readthedocs/search/faceted_search.py

```diff
@@ -96,53 +95,112 @@ class PageSearchBase(RTDFacetedSearch):
-    operators = ['and']
+        # match query for the title (of the page) field.
+        match_title_query = Match(title=query)
+
+        # nested query for search in sections
+        sections_nested_query = self.generate_nested_query(
+            query=query,
+            path='sections',
+            fields=self.section_fields,
+            inner_hits={
+                'highlight': {
+                    'fields': {
+                        'sections.title': {},
+                        'sections.content': {},
+                    }
+                }
+            }
+        )
+
+        # nested query for search in domains
+        domains_nested_query = self.generate_nested_query(
+            query=query,
+            path='domains',
+            fields=self.domain_fields,
+            inner_hits={
+                'highlight': {
+                    'fields': {
+                        'domains.type_display': {},
+                        'domains.doc_display': {},
+                        'domains.name': {},
+                        'domains.display_name': {},
+                    }
+                }
+            }
+        )
+
+        final_query = Bool(should=[
+            match_title_query,
+            sections_nested_query,
+            domains_nested_query,
+        ])
+
+        search = search.query(final_query)
+        return search
 
 
-class PageSearch(SettingsOverrideObject):
+    def generate_nested_query(self, query, path, fields, inner_hits):
+        """Generate a nested query with passed parameters."""
+        queries = []
 
-    """
-    Allow this class to be overridden based on CLASS_OVERRIDES setting.
+        for operator in self.operators:
+            query_string = SimpleQueryString(
+                query=query,
+                fields=fields,
+                default_operator=operator
+            )
+            queries.append(query_string)
 
-    This is primary used on the .com to adjust how we filter our search queries
-    """
+        bool_query = Bool(should=queries)
 
-    _default_class = PageSearchBase
+        nested_query = Nested(
+            path=path,
+            inner_hits=inner_hits,
+            query=bool_query
+        )
+        return nested_query
 
 
-class ProjectSearch(SettingsOverrideObject):
+class PageSearch(SettingsOverrideObject):
 
     """
     Allow this class to be overridden based on CLASS_OVERRIDES setting.
 
     This is primary used on the .com to adjust how we filter our search queries
     """
 
-    _default_class = ProjectSearchBase
+    _default_class = PageSearchBase
 
 
-class DomainSearch(SettingsOverrideObject):
+class ProjectSearch(SettingsOverrideObject):
 
     """
     Allow this class to be overridden based on CLASS_OVERRIDES setting.
 
     This is primary used on the .com to adjust how we filter our search queries
     """
 
-    _default_class = DomainSearchBase
+    _default_class = ProjectSearchBase
 
 
 class AllSearch(RTDFacetedSearch):
```

Comment:
I think we can probably remove this for now, since I think nesting is a better approach to this problem.

Reply:
Sounds good.
I have removed it.